core/src/apps/cardano/sign_tx.py

```diff
@@ -125,117 +109,82 @@ async def sign_tx(ctx, msg, keychain: seed.Keychain):
-            )
+            node = self.keychain.derive(input.address_n)
+            message = b"\x58\x20" + tx_aux_hash
+
             signature = ed25519.sign_ext(
                 node.private_key(), node.private_key_ext(), message
             )
-            extended_public_key = (
-                remove_ed25519_prefix(node.public_key()) + node.chain_code()
-            )
-            witnesses.append(
-                [
-                    (input.type or 0),
-                    cbor.Tagged(24, cbor.encode([extended_public_key, signature])),
-                ]
-            )
-
-        return witnesses
-
-    @staticmethod
-    def compute_fee(input_coins_sum: int, outgoing_coins: list, change_coins: list):
-        outgoing_coins_sum = sum(outgoing_coins)
-        change_coins_sum = sum(change_coins)
-
-        return input_coins_sum - outgoing_coins_sum - change_coins_sum
 
-    def serialise_tx(self):
+            public_key = remove_ed25519_prefix(node.public_key())
 
-        self._process_outputs()
+            # todo: GK - verify this works with Byron inputs (after IOHK confirmation/testnet support)
+            witnesses.append([public_key, signature])
 
-        inputs_cbor = []
-        for input in self.inputs:
-            inputs_cbor.append(
-                [
-                    (input.type or 0),
-                    cbor.Tagged(24, cbor.encode([input.prev_hash, input.prev_index])),
-                ]
-            )
-
-        inputs_cbor = cbor.IndefiniteLengthArray(inputs_cbor)
-
-        outputs_cbor = []
-        for index, address in enumerate(self.output_addresses):
-            outputs_cbor.append(
-                [cbor.Raw(base58.decode(address)), self.outgoing_coins[index]]
-            )
+        return witnesses
 
-        for index, address in enumerate(self.change_addresses):
-            outputs_cbor.append(
-                [cbor.Raw(base58.decode(address)), self.change_coins[index]]
-            )
+    def serialise_tx(self) -> Tuple[bytes, bytes]:
+        inputs_for_cbor = self._build_inputs()
+        outputs_for_cbor = self._build_outputs()
 
-        outputs_cbor = cbor.IndefiniteLengthArray(outputs_cbor)
+        tx_body = {0: inputs_for_cbor, 1: outputs_for_cbor, 2: self.fee, 3: self.ttl}
 
-        tx_aux_cbor = [inputs_cbor, outputs_cbor, self.attributes]
-        tx_hash = hashlib.blake2b(data=cbor.encode(tx_aux_cbor), outlen=32).digest()
+        tx_body_cbor = cbor.encode(tx_body)
+        tx_hash = hashlib.blake2b(data=tx_body_cbor, outlen=32).digest()
 
-        witnesses = self._build_witnesses(tx_hash)
-        tx_body = cbor.encode([tx_aux_cbor, witnesses])
+        witnesses_for_cbor = self._build_witnesses(tx_hash)
+        witnesses = {0: witnesses_for_cbor}
 
-        self.fee = self.compute_fee(
-            self.input_coins_sum, self.outgoing_coins, self.change_coins
-        )
+        tx = cbor.encode([tx_body, witnesses, None])
```

Comment:
I would name this serialised_tx

Reply:
I'll rename it.

I need to build the `tx_body` beforehand in order to hash it for signing. So then only witnesses would remain to be added as you describe, but I am not sure it's worth it.

Reply:
I see, ok nevermind